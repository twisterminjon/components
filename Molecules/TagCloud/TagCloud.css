.tagcloud {
  position: relative;
  list-style-type: none;
}

.tagcloud--vertical {
  /* display: block; */
}

.tagcloud--disabled {
  pointer-events: none;
}
