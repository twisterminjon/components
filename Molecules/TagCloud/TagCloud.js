import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Tag from '../Tag/Tag'
import './TagCloud.css'

export default class TagCloud extends Component {
  static propTypes = {
    /** Can show a loader on the tags */
    loading: PropTypes.bool,
    tags: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        name: PropTypes.string.isRequired,
        isActive: PropTypes.bool,
      })
    ).isRequired,
    vertical: PropTypes.bool,

    /** Called after tag is deleted */
    onTagDelete: PropTypes.func.isRequired,

    /** React view for individual tags */
    TagView: PropTypes.any, // FIXME: Define better type
  }
  static defaultProps = {
    loading: false,
    vertical: true,
    TagView: ({ tagObject, ...props }) => {
      const { onTagDelete, vertical } = props
      const orientationClass = vertical ? 'tagcloud--vertical' : ''

      return (
        <Tag
          label={tagObject.name}
          key={tagObject.id}
          id={tagObject.id}
          inActive={!tagObject.isActive}
          onDelete={(id, name) => onTagDelete(id, name)}
          className={orientationClass}
          fluid={vertical ? true : false}
          style={{ marginBottom: 6, marginRight: vertical ? 0 : 6 }}
        />
      )
    },
  }

  render() {
    const { loading, tags, TagView, 'data-testid': dataTestId, ...rest } = this.props

    const loadingClass = loading ? 'button-spinner tagcloud--disabled' : ''

    return (
      <ul className={`tagcloud ${loadingClass}`.trim()} data-testid={dataTestId}>
        {tags.map(tagObject => {
          const { id, surveySessionId } = tagObject

          // Preference surveySessionId over id for key, if available
          const key = surveySessionId || id

          return <TagView key={key} tagObject={tagObject} {...rest} />
        })}
      </ul>
    )
  }
}
