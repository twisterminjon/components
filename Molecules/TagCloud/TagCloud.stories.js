import React from 'react'

import { storiesOf } from '@storybook/react'
import { withKnobs, boolean } from '@storybook/addon-knobs'
import centered from '@storybook/addon-centered/react'

import TagCloud from './TagCloud'

const stories = storiesOf('Provider/v1/Molecules', module)
stories.addDecorator(withKnobs)

stories.addDecorator(centered).add('TagCloud', () => {
  const tags = [
    { id: '1', name: 'Fantastic Four', isActive: true },
    { id: '2', name: 'X-Men', isActive: true },
    { id: '3', name: 'Justice League of America', isActive: true },
    { id: '4', name: 'The Avengers', isActive: false },
    { id: '5', name: 'Batman/Superman', isActive: true },
  ]

  const verticalVal = boolean('vertical', false)
  const loadingVal = boolean('loading', false)

  return (
    <div style={{ width: 400 }}>
      <TagCloud
        loading={loadingVal}
        tags={tags}
        vertical={verticalVal}
        onTagDelete={(id, name) => {
          alert(`delete tag ${id}-${name}`)
        }}
      />
    </div>
  )
})
