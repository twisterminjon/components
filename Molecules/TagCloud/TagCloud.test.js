import React from 'react'
import TagCloud from './TagCloud'

describe('TagCloud', () => {
  const tagList = [
    { id: '1', name: 'Fantastic Four' },
    { id: '2', name: 'X-Men' },
  ]
  const mockFun = jest.fn()

  it('matches the snapshot', () => {
    const wrapper = window.shallow(<TagCloud tags={tagList} onTagDelete={mockFun} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 2 tags', () => {
    const wrapper = window.shallow(<TagCloud tags={tagList} onTagDelete={mockFun} />)
    expect(wrapper.children()).toHaveLength(2)
  })

  it('clicking on delete on a tag should call the delete function', () => {
    const wrapper = window.mount(<TagCloud tags={tagList} onTagDelete={mockFun} />)

    wrapper
      .find('[data-testid="tag-Fantastic Four-delete-button"]')
      .last() // Use last rendered component as the virtual DOM shows React components in addition to actual DOM components
      .simulate('click')

    expect(mockFun).toHaveBeenCalledWith('1', 'Fantastic Four')
  })
})
