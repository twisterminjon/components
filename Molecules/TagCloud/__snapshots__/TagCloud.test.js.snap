// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagCloud matches the snapshot 1`] = `
<ul
  className="tagcloud"
>
  <TagView
    key="1"
    onTagDelete={[MockFunction]}
    tagObject={
      Object {
        "id": "1",
        "name": "Fantastic Four",
      }
    }
    vertical={true}
  />
  <TagView
    key="2"
    onTagDelete={[MockFunction]}
    tagObject={
      Object {
        "id": "2",
        "name": "X-Men",
      }
    }
    vertical={true}
  />
</ul>
`;
